### ISSUE TITLE
Support for Pydantic Model as Signature InputField type

### ISSUE THREAD

**Issue by rohitgarud** [stanfordnlp/dspy#1904]
Currently, only Signature OutputField can have Pydantic Model as a type. Adding it to InputField does not reflect in the System prompt

**Comment by okhat** [comment]
Hey @rohitgarud , it works though, right? It just doesn't describe the schema. We don't need to describe the schema of the input, since we show the input value to the model.

**Comment by rohitgarud** [comment]
Yes, @okhat  it works, I was wondering if adding the description of the fields in the input Pydantic model can help use the provided information better

**Comment by okhat** [comment]
It was an analytic decision. Why should we show the schema to the model, if it sees the actual structure + values anyway?

However, I'm interested in whether there _are_ edge cases where showing the input schema is important.

**Comment by rohitgarud** [comment]
What would you prefer: 
1. multiple InputFields with descriptions, say 5-6 
2. single InputField with Pydantic model and show the schema using a custom adapter

**Comment by okhat** [comment]
Of course the first one :D

**Comment by rohitgarud** [comment]
In that case, what happens if the data for any InputField is optional and missing?.. I will be testing this, but I wanted to get your intuition on this

**Comment by rohitgarud** [comment]
> What would you prefer:
> 
> 1. multiple (say 5-6) InputFields with descriptions
> 2. single InputField with Pydantic model and show the schema using a custom adapter

Going against your suggestion, went with option 2 😄 by doing this in the custom adapter (a little hacky but will be improving later):

```python
def field_metadata(field_name, field_info):
        type_ = field_info.annotation

        if type_ is str:
            desc = ""
        elif type_ is bool:
            desc = "must be True or False"
        elif type_ in (int, float):
            desc = f"must be a single {type_.__name__} value"
        elif inspect.isclass(type_) and issubclass(type_, enum.Enum):
            desc = f"must be one of: {'; '.join(type_.__members__)}"
        elif hasattr(type_, "__origin__") and type_.__origin__ is Literal:
            desc = f"must be one of: {'; '.join([str(x) for x in type_.__args__])}"  # noqa: E501
        else:
            desc = (
                "must be pareseable according to the following JSON schema: "
            )
            processed_schema = ProcessSchema(
                schema=TypeAdapter(type_).json_schema()
            ).transform_schema()
            desc += processed_schema
        desc = (
            (" " * 8) + f"# note: the value you produce {desc}" if desc else ""
        )

        if get_dspy_field_type(field_info) == "input":
            desc = desc.replace(
                "# note: the value you produce must be", "#note: input will be"
            )
            desc = desc.replace("pareseable according to the", "having")

        return f"{{{field_name}}}{desc}"
```

Rationale:
- Fields and descriptions of inputs and outputs stay at a single location inside models, keeping the program cleaner
- Can use nested pedantic models for input
- As I am [processing the schema ](https://gist.github.com/rohitgarud/80bdcd30b65c154e07f343055f95898e) before injecting into the system prompt using [custom adapter](https://gist.github.com/rohitgarud/eb60c095a53cf5303fb3ae07b98e268b) the increase in number of tokens is manageable 

Please let me know your thoughts on this

**Comment by thomasahle** [comment]
Sometimes a codebase may already have a bunch of pydantic types it uses. It's inconvenient to "unpack" them into a signature, rather than just using them directly. So maybe some automatic "unpacking formatter" is not a bad idea.

**Comment by rohitgarud** [comment]
Thank you @thomasahle, I would really appreciate if you can review [this approach](https://gist.github.com/rohitgarud/80bdcd30b65c154e07f343055f95898e) for unpacking formatter. Works really well with even smaller models. I think we can still use the Structured Output feature from the providers. Usage is shown above in `ProcessSchema`

**Comment by davidberenstein1957** [comment]
Does something like this work? https://github.com/davidberenstein1957/dspydantic
